### PR TITLE
Various wallet issues

### LINKED ISSUE #145: Split transaction low fee issue
A transaction that includes >33 receipients gets split in multiple transactions:

    No payment id is included with this transaction. Is this okay?  (Y/Yes/N/No): y
    
    Transaction 1/2:
    Spending from address index 0
    
    Transaction 2/2:
    Spending from address index 0
    Sending 693.000000000.  Your transaction needs to be split into 2 transactions.  This will result in a transaction fee being applied to each transaction, for a total fee of 0.046000000
The second transaction fails to go through due to low fees, unless the fee is manually set to above normal:

    Transaction successfully submitted, transaction <72ee6aa7daabd25940479630b3aacb44ff628f9df6f39968f58307bb45d5e681>
    You can check its status by using the `show_transfers` command.
    Error: transaction <8cf49ed214340754c893033c88c17dddad6aff5ab39e4968f9ed53f521287101> was rejected by daemon with status: Failed
    Error: Reason: fee too low

### PR DESCRIPTION
**Pull request by fireice-uk**
- Optimise `pick_preferred_rct_inputs`
- Fix #145 
- Implement seamless max out handling

### CODE REVIEW

File: src/wallet/wallet2.cpp [hunk 1]
```diff
@@ -6858,8 +6858,12 @@ std::vector<wallet2::pending_tx> wallet2::create_transactions_2(std::vector<cryp
 		}
 		else
 		{
-			while(!dsts.empty() && dsts[0].amount <= available_amount && estimate_rct_tx_size(tx.selected_transfers.size(), fake_outs_count, tx.dsts.size(), bulletproof) < TX_SIZE_TARGET(upper_transaction_size_limit))
+			while(!dsts.empty() && dsts[0].amount <= available_amount && 
+				estimate_rct_tx_size(tx.selected_transfers.size(), fake_outs_count, tx.dsts.size(), bulletproof) < TX_SIZE_TARGET(upper_transaction_size_limit))
 			{
+				if(bulletproof && tx.dsts.size() >= cryptonote::common_config::BULLETPROOF_MAX_OUTPUTS-1)
```
Comment (psychocrypt) [inline src/wallet/wallet2.cpp]: please recheck if this must be `>` and later on line `6879` `<=`

File: src/wallet/wallet2.cpp [hunk 2]
```diff
@@ -6869,13 +6873,17 @@ std::vector<wallet2::pending_tx> wallet2::create_transactions_2(std::vector<cryp
...
-			if(available_amount > 0 && !dsts.empty() && estimate_rct_tx_size(tx.selected_transfers.size(), fake_outs_count, tx.dsts.size(), bulletproof) < TX_SIZE_TARGET(upper_transaction_size_limit))
+			if(available_amount > 0 && !dsts.empty() && 
+				estimate_rct_tx_size(tx.selected_transfers.size(), fake_outs_count, tx.dsts.size(), bulletproof) < TX_SIZE_TARGET(upper_transaction_size_limit))
 			{
-				// we can partially fill that destination
-				LOG_PRINT_L2("We can partially pay " << get_public_address_as_str(m_nettype, dsts[0].is_subaddress, dsts[0].addr) << " for " << print_money(available_amount) << "/" << print_money(dsts[0].amount));
-				tx.add(dsts[0].addr, dsts[0].is_subaddress, available_amount, original_output_index, m_merge_destinations);
-				dsts[0].amount -= available_amount;
-				available_amount = 0;
+				if(!bulletproof || tx.dsts.size() < cryptonote::common_config::BULLETPROOF_MAX_OUTPUTS-1)
```
Comment (psychocrypt) [inline src/wallet/wallet2.cpp]: Must this be `<=` because we can have 15 outputs.

File: src/wallet/wallet2.cpp (hunk 1 above)
Comment (fireice-uk) [reply src/wallet/wallet2.cpp]: Nope. We need one free slot for change.

File: src/wallet/wallet2.cpp (hunk 2 above)
Comment (fireice-uk) [reply src/wallet/wallet2.cpp]: Same.

File: src/wallet/wallet2.cpp [hunk 3]
```diff
@@ -6393,56 +6393,99 @@ void wallet2::transfer_selected_rct(std::vector<cryptonote::tx_destination_entry
...
-		{
-			LOG_PRINT_L2("Considering input " << i << ", " << print_money(td.amount()));
-			for(size_t j = i + 1; j < m_transfers.size(); ++j)
-			{
-				const transfer_details &td2 = m_transfers[j];
-				if(!td2.m_spent && !td.m_key_image_partial && td2.is_rct() && td.amount() + td2.amount() >= needed_money && is_transfer_unlocked(td2) && td2.m_subaddr_index == td.m_subaddr_index)
+	for(size_t i = 0; i < pick_list.size(); i++)
+	{
+		LOG_PRINT_L2("Considering input " << pick_list[i].idx << ", " << print_money(pick_list[i].amount));
+		for(size_t j = pick_list.size(); --j > i+1;)
```
[the remark this answers is not in the document]
Comment (fireice-uk) [reply src/wallet/wallet2.cpp]: I was thinking of `j--`, will update

File: src/wallet/wallet2.cpp [hunk 4]
```diff
@@ -6582,6 +6582,7 @@ std::vector<wallet2::pending_tx> wallet2::create_transactions_2(std::vector<cryp
 		cryptonote::transaction tx;
 		pending_tx ptx;
 		size_t bytes;
+		uint64_t fee;
```
[the remark this answers is not in the document]
Comment (fireice-uk) [reply src/wallet/wallet2.cpp]: Yes, will do

File: src/wallet/wallet2.cpp (hunk 1 above)
Comment (psychocrypt) [reply src/wallet/wallet2.cpp]: But since you already reduced the max outputs by one and use `>=` this means we use only 15 outputs.

Comment (fireice-uk) [reply src/wallet/wallet2.cpp]: That's correct. You need one free slot for change.

File: src/wallet/wallet2.cpp [hunk 5]
```diff
@@ -6393,56 +6393,99 @@ void wallet2::transfer_selected_rct(std::vector<cryptonote::tx_destination_entry
...
-		{
-			LOG_PRINT_L2("Considering input " << i << ", " << print_money(td.amount()));
-			for(size_t j = i + 1; j < m_transfers.size(); ++j)
-			{
-				const transfer_details &td2 = m_transfers[j];
-				if(!td2.m_spent && !td.m_key_image_partial && td2.is_rct() && td.amount() + td2.amount() >= needed_money && is_transfer_unlocked(td2) && td2.m_subaddr_index == td.m_subaddr_index)
+	for(size_t i = 0; i < pick_list.size(); i++)
+	{
+		LOG_PRINT_L2("Considering input " << pick_list[i].idx << ", " << print_money(pick_list[i].amount));
+		for(size_t j = pick_list.size(); j-- > i+1;)
```
Comment (psychocrypt) [inline src/wallet/wallet2.cpp]: now it is a out of memory acces because j is in the first round one element behind the last element in the container.

Comment (psychocrypt) [reply src/wallet/wallet2.cpp]: All fine j is decremented after the comparison

**Pull request merged**